temporal/fx.go

```diff
@@ -481,46 +481,94 @@ func ApplyClusterMetadataConfigProvider(
 	}
 	defer clusterMetadataManager.Close()
 
-	applied, err := clusterMetadataManager.SaveClusterMetadata(
-		&persistence.SaveClusterMetadataRequest{
-			ClusterMetadata: persistencespb.ClusterMetadata{
-				HistoryShardCount: config.Persistence.NumHistoryShards,
-				ClusterName:       config.ClusterMetadata.CurrentClusterName,
-				ClusterId:         uuid.New(),
-			}})
-	if err != nil {
-		logger.Warn("Failed to save cluster metadata.", tag.Error(err))
-	}
-	if applied {
-		logger.Info("Successfully saved cluster metadata.")
-		return config.ClusterMetadata, config.Persistence, nil
-	}
+	clusterData := config.ClusterMetadata
+	for clusterName, clusterInfo := range clusterData.ClusterInformation {
+		var clusterId = ""
+		if clusterName == clusterData.CurrentClusterName {
+			// Only set current cluster Id as we don't know the remote cluster Id.
+			clusterId = uuid.New()
+		}
+		// We assume the existing remote cluster info is correct.
+		applied, err := clusterMetadataManager.SaveClusterMetadata(
+			&persistence.SaveClusterMetadataRequest{
+				ClusterMetadata: persistencespb.ClusterMetadata{
+					HistoryShardCount:        config.Persistence.NumHistoryShards,
+					ClusterName:              clusterName,
+					ClusterId:                clusterId,
+					ClusterAddress:           clusterInfo.RPCAddress,
+					FailoverVersionIncrement: clusterData.FailoverVersionIncrement,
+					InitialFailoverVersion:   clusterInfo.InitialFailoverVersion,
+					IsGlobalNamespaceEnabled: clusterData.EnableGlobalNamespace,
+					IsConnectionEnabled:      clusterInfo.Enabled,
+				}})
+		if err != nil {
+			logger.Warn("Failed to save cluster metadata.", tag.Error(err), tag.ClusterName(clusterName))
+		}
 
-	resp, err := clusterMetadataManager.GetClusterMetadata()
-	if err != nil {
-		return config.ClusterMetadata, config.Persistence, fmt.Errorf("error while fetching cluster metadata: %w", err)
-	}
+		if applied {
+			logger.Info("Successfully saved cluster metadata.", tag.ClusterName(clusterName))
+			continue
+		}
 
-	if config.ClusterMetadata.CurrentClusterName != resp.ClusterName {
-		logger.Error(
-			mismatchLogMessage,
-			tag.Key("clusterMetadata.currentClusterName"),
-			tag.IgnoredValue(config.ClusterMetadata.CurrentClusterName),
-			tag.Value(resp.ClusterName))
+		resp, err := clusterMetadataManager.GetCurrentClusterMetadata()
+		if err != nil {
+			return config.ClusterMetadata,
+				config.Persistence,
+				fmt.Errorf("error while fetching metadata from cluster %s: %w", clusterName, err)
+		}
 
-		config.ClusterMetadata.CurrentClusterName = resp.ClusterName
-	}
+		if clusterName == clusterData.CurrentClusterName {
+			if config.ClusterMetadata.CurrentClusterName != resp.ClusterName {
```

Comment:
how could this happen? the resp is from GetCurrentClusterMetadata() which load from DB for currentClusterName.

Reply:
If the current cluster name in config file is changed, the GetCurrentClusterMetadata() would load for the change cluster name, so the returned response's ClusterName would always equal to the currentClusterName from config.

Reply:
Correct. This check is pointless with the new change. I am going to remove it.